File: app/src/main/java/pl/trikimusic/controller/domain/model/ControlModels.kt
package pl.trikimusic.controller.domain.model

import kotlinx.serialization.Serializable

@Serializable
enum class ButtonClickType(
    val clickCount: Int,
    val displayName: String,
) {
    SINGLE(1, "Jeden klik"),
    DOUBLE(2, "Dwa kliknięcia"),
    TRIPLE(3, "Trzy kliknięcia"),
}

data class ButtonClickEvent(
    val type: ButtonClickType,
    val timestampNanos: Long,
)

@Serializable
enum class MediaAction(val displayName: String) {
    PLAY("Odtwórz"),
    PAUSE("Wstrzymaj"),
    PLAY_PAUSE("Odtwórz / wstrzymaj"),
    NEXT("Następny utwór"),
    PREVIOUS("Poprzedni utwór"),
    LIKE("Polub utwór"),
    DISLIKE("Odrzuć utwór"),
    VOLUME_UP("Głośniej"),
    VOLUME_DOWN("Ciszej"),
    MUTE("Wycisz"),
    UNMUTE("Wyłącz wyciszenie"),
    STOP("Zatrzymaj"),
    NONE("Brak akcji"),
}

@Serializable
data class ButtonMapping(
    val click: ButtonClickType,
    val action: MediaAction,
)

fun defaultButtonMappings(): List<ButtonMapping> = listOf(
    ButtonMapping(ButtonClickType.SINGLE, MediaAction.PLAY_PAUSE),
    ButtonMapping(ButtonClickType.DOUBLE, MediaAction.NEXT),
    ButtonMapping(ButtonClickType.TRIPLE, MediaAction.PREVIOUS),
)

@Serializable
data class ControlProfile(
    val id: String,
    val name: String,
    val builtIn: Boolean = false,
    val buttonMappings: List<ButtonMapping> = defaultButtonMappings(),
) {
    fun actionFor(click: ButtonClickType): MediaAction =
        buttonMappings.firstOrNull { it.click == click }?.action ?: MediaAction.NONE
}

File: windows/TrikiMusicController.Windows/Models/ControlModels.cs
namespace TrikiMusicController_Windows.Models;

public enum ButtonClickType
{
    Single = 1,
    Double = 2,
    Triple = 3,
}

public enum MediaAction
{
    Play,
    Pause,
    PlayPause,
    Next,
    Previous,
    Like,
    Dislike,
    VolumeUp,
    VolumeDown,
    Mute,
    Unmute,
    Stop,
    None,
}

public static class MediaActionNames
{
    public static string DisplayName(this MediaAction action) => action switch
    {
        MediaAction.Play => "Odtwórz",
        MediaAction.Pause => "Wstrzymaj",
        MediaAction.PlayPause => "Odtwórz / wstrzymaj",
        MediaAction.Next => "Następny utwór",
        MediaAction.Previous => "Poprzedni utwór",
        MediaAction.Like => "Polub utwór",
        MediaAction.Dislike => "Odrzuć utwór",
        MediaAction.VolumeUp => "Głośniej",
        MediaAction.VolumeDown => "Ciszej",
        MediaAction.Mute => "Wycisz",
        MediaAction.Unmute => "Wyłącz wyciszenie",
        MediaAction.Stop => "Zatrzymaj",
        _ => "Brak akcji",
    };
}

public sealed record ButtonClickEvent(ButtonClickType Type, long TimestampNanos);

public sealed record MediaActionOption(MediaAction Action, string DisplayName)
{
    public string Name => DisplayName;
    public override string ToString() => DisplayName;
}

public sealed record ThemeOption(string Value, string DisplayName)
{
    public string Name => DisplayName;
    public override string ToString() => DisplayName;
}

public sealed record MultiDeviceArbitrationOption(MultiDeviceArbitrationMode Mode, string DisplayName)
{
    public string Name => DisplayName;
    public override string ToString() => DisplayName;
}

public enum TrikiButtonProtocolMode
{
    Unknown,
    ButtonFlag,
    SequenceCounter,
}

public sealed record VolumeControlResult(
    MediaAction? Action,
    bool SensorValid,
    bool WithinTiltRange,
    bool AccelerationStable,
    bool TiltStable,
    float StabilizationProgress,
    bool Active,
    float TiltDegrees,
    float GyroscopeZDps);

public enum RatingGestureAction
{
    Like,
    Dislike,
}

public enum RotationGestureDirection
{
    Right,
    Left,
}

public static class RotationGestureDirectionActions
{
    public static MediaAction ToInvertedCapsuleNavigationAction(this RotationGestureDirection direction) => direction switch
    {
        RotationGestureDirection.Left => MediaAction.Next,
        RotationGestureDirection.Right => MediaAction.Previous,
        _ => MediaAction.None,
    };
}

public enum HoldGesturePhase
{
    Idle,
    Holding,
    Ready,
    Tracking,
    Completing,
    Rearming,
    Triggered,
}

public sealed record HoldArcGestureResult(
    RatingGestureAction? Action,
    RatingGestureAction? Direction,
    HoldGesturePhase Phase,
    float HoldProgress,
    bool FaceDown,
    float EstimatedHorizontalDisplacementMeters,
    float EstimatedArcDepthMeters);

public sealed record FullRotationGestureResult(
    bool Triggered,
    RotationGestureDirection? Direction,
    HoldGesturePhase Phase,
    float StabilizationProgress,
    bool FaceDown,
    float EstimatedRotationDegrees,
    float GyroscopeZDps);
